{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a305d697",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8a46bff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess(data_mat, meta_data, IDCol, covar_l = []):\n",
    "    # remove samples with all zeros\n",
    "    data_mat = data_mat.loc[~(data_mat==0).all(axis=1)]\n",
    "    kept_samples = data_mat.index\n",
    "    meta_data = meta_data[meta_data[IDCol].isin(kept_samples)]\n",
    "    # remove features with all zeros\n",
    "    col_names = list(data_mat.columns)\n",
    "    col_sums = data_mat.sum(axis = 1)\n",
    "    removable_feature_names = [col_names[index] for index, col_sum in enumerate(col_sums) if col_sum==0]\n",
    "    data_mat.drop(removable_feature_names, axis=1, inplace=True)\n",
    "    # for each covar used, remove samples with missing values\n",
    "    for covar in covar_l:\n",
    "        meta_data = meta_data[meta_data[covar].notna()]\n",
    "    data_mat = data_mat.loc[meta_data[IDCol]]\n",
    "    # after cleaning samples, remove features with all zeros again\n",
    "    col_names = list(data_mat.columns)\n",
    "    col_sums = data_mat.sum(axis = 1)\n",
    "    removable_feature_names = [col_names[index] for index, col_sum in enumerate(col_sums) if col_sum==0]\n",
    "    data_mat.drop(removable_feature_names, axis=1, inplace=True)\n",
    "    return data_mat, meta_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1c04229b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_data_phyloseq(address_directory, output_root = False, id = 'Sam_id', covar_l = [], relab = True):\n",
    "    ### CuratedMetagenomicsDataset provides way more metadata in a congestible manner\n",
    "    cur_dir_names = os.listdir(address_directory)\n",
    "    address_X = address_directory + '/'+ [result for result in cur_dir_names if \"otu_table_\" in result][0]\n",
    "    address_Y = address_directory + '/'+ [result for result in cur_dir_names if \"sample_table_\" in result][0]\n",
    "    data_mat = pd.read_csv(address_X, index_col=0)\n",
    "    meta_data = pd.read_csv(address_Y, index_col=0)\n",
    "    meta_data[id] = list(meta_data.index)\n",
    "    data_mat, meta_data = preprocess(data_mat.T, meta_data, id, covar_l)\n",
    "    \n",
    "    if not relab:\n",
    "        # get taxa names\n",
    "        taxa_df = pd.read_csv(address_directory + '/'+ [result for result in cur_dir_names if \"tax_table_\" in result][0], index_col=0)\n",
    "        if 'ta6' in taxa_df.columns:\n",
    "            taxa_names = list(taxa_df['ta6'])\n",
    "        elif 'genus' in taxa_df.columns:\n",
    "            taxa_names = list(taxa_df['genus'])\n",
    "        else:\n",
    "            taxa_names = list(taxa_df['Genus'])\n",
    "        data_mat.columns = taxa_names\n",
    "        \n",
    "    # TODO: ensure that the sample ids are correctly aligned in metadata and count_table\n",
    "    data_mat_ids = list(data_mat.index)\n",
    "    meta_data_ids = list(meta_data.index)\n",
    "    intersection_ids = list(set(meta_data_ids).intersection(data_mat_ids))\n",
    "\n",
    "    # drop rows where indexes are not overlapping\n",
    "    data_mat_non_intersecting = [id for id in data_mat_ids if id not in intersection_ids]\n",
    "    data_mat = data_mat.drop(data_mat_non_intersecting)\n",
    "    meta_data_non_intersecting = [id for id in meta_data_ids if id not in intersection_ids]\n",
    "    meta_data = meta_data.drop(meta_data_non_intersecting)\n",
    "    data_mat = data_mat.reindex(intersection_ids)\n",
    "    meta_data = meta_data.reindex(intersection_ids)\n",
    "\n",
    "    # drop taxa where all values are 0\n",
    "    data_mat = data_mat.loc[:, (data_mat != 0).any(axis=0)]\n",
    "    # data_mat = data_mat.loc[(data_mat != 0).any(axis=1), :]\n",
    "\n",
    "    # convert data_mat to relative abundance for each sample\n",
    "    if relab:\n",
    "        data_mat = data_mat.div(data_mat.sum(axis=1), axis=0) # divide by row sum\n",
    "\n",
    "    # save stuff if needed\n",
    "    if output_root != False:\n",
    "        data_mat.to_csv(output_root+\"_count_data.csv\") # this is actually not count, but relative abundance\n",
    "        meta_data.to_csv(output_root+\"_meta_data.csv\", index=False)\n",
    "    return data_mat, meta_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ea801463",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_results_from_benchmarked_methods(address_X, address_Y):\n",
    "    data_mat = pd.read_csv(address_X, index_col=0)\n",
    "    meta_data = pd.read_csv(address_Y)\n",
    "\n",
    "    # check for missing values and fill out with mean\n",
    "    if data_mat.isnull().values.any():\n",
    "        print(\"fill out nans\")\n",
    "        data_mat = data_mat.fillna(data_mat.mean())\n",
    "        # remove columns (taxa) filled with NAs\n",
    "        data_mat = data_mat.dropna(axis=1, how='all')\n",
    "\n",
    "    return data_mat, meta_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fa18b016",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_complete_confounding(meta_data, batch_var, bio_var, output_root = ''):\n",
    "    # make a pandas dataframe where rows are batches whereas columns are the bio_var options\n",
    "    # each entry is the number of samples in that batch with that bio_var option\n",
    "    # if there is a batch with only one bio_var option, then it is a complete confounder\n",
    "    print(meta_data)\n",
    "    # get the list of batches\n",
    "    batch_l = list(meta_data[batch_var])\n",
    "    # batch_l = [x for x in batch_l if str(x) != 'nan']\n",
    "    batch_l = list(np.unique(batch_l))\n",
    "\n",
    "    # get the list of bio_var options\n",
    "    bio_var_l = list(meta_data[bio_var])\n",
    "    # bio_var_l = [x for x in bio_var_l if str(x) != 'nan']\n",
    "    bio_var_l = list(np.unique(bio_var_l))\n",
    "\n",
    "    # generate a dataframe\n",
    "    df = pd.DataFrame(columns=bio_var_l, index=batch_l)\n",
    "    for batch in batch_l:\n",
    "        for bio_var_val in bio_var_l:\n",
    "            df.loc[batch, bio_var_val] = len(meta_data.loc[(meta_data[batch_var]==batch) & (meta_data[bio_var]==bio_var_val)])\n",
    "    \n",
    "    if output_root != '':\n",
    "        df.to_csv(output_root+\"_complete_confounding.csv\")\n",
    "    print(df)\n",
    "    # check if there is a batch with only one bio_var option\n",
    "    for batch in batch_l:\n",
    "        if len(df.loc[batch].unique())==1:\n",
    "            print(\"batch\", batch, \"is a complete confounder\")\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "457a0175",
   "metadata": {},
   "outputs": [],
   "source": [
    "# overall_path = '/athena/linglab/scratch/chf4012/mic_bc_benchmark/data'\n",
    "# # autism_2_microbiomeHD\n",
    "# data_mat, meta_data = preprocess_data_phyloseq(f'{overall_path}/pruned_autism_2_microbiomeHD', f'{overall_path}/cleaned_data/autism_2_microbiomeHD/autism_2_microbiomeHD', id = 'Sam_id', covar_l = [], relab = False)\n",
    "# data_mat, meta_data = load_results_from_benchmarked_methods(f'{overall_path}/cleaned_data/autism_2_microbiomeHD/autism_2_microbiomeHD_count_data.csv', f'{overall_path}/cleaned_data/autism_2_microbiomeHD/autism_2_microbiomeHD_meta_data.csv')\n",
    "# check_complete_confounding(meta_data, 'Dataset', 'DiseaseState', f'{overall_path}/cleaned_data/autism_2_microbiomeHD/autism_2_microbiomeHD')\n",
    "\n",
    "# # cdi_3_microbiomeHD\n",
    "# data_mat, meta_data = preprocess_data_phyloseq(f'{overall_path}/pruned_cdi_3_microbiomeHD', f'{overall_path}/cleaned_data/cdi_3_microbiomeHD/cdi_3_microbiomeHD', id = 'Sam_id', covar_l = [], relab = False)\n",
    "# data_mat, meta_data = load_results_from_benchmarked_methods(f'{overall_path}/cleaned_data/cdi_3_microbiomeHD/cdi_3_microbiomeHD_count_data.csv', f'{overall_path}/cleaned_data/cdi_3_microbiomeHD/cdi_3_microbiomeHD_meta_data.csv')\n",
    "# check_complete_confounding(meta_data, 'Dataset', 'DiseaseState', f'{overall_path}/cleaned_data/cdi_3_microbiomeHD/cdi_3_microbiomeHD')\n",
    "\n",
    "# # ibd_3_CMD\n",
    "# data_mat, meta_data = preprocess_data_phyloseq(f'{overall_path}/pruned_ibd_3_CMD', f'{overall_path}/cleaned_data/ibd_3_CMD/ibd_3_CMD', id = 'Sam_id', covar_l = ['disease', 'gender', 'age_category'])\n",
    "# data_mat, meta_data = load_results_from_benchmarked_methods(f'{overall_path}/cleaned_data/ibd_3_CMD/ibd_3_CMD_count_data.csv', f'{overall_path}/cleaned_data/ibd_3_CMD/ibd_3_CMD_meta_data.csv')\n",
    "# check_complete_confounding(meta_data, \"study_name\", \"disease\", f'{overall_path}/cleaned_data/ibd_3_CMD/ibd_3_CMD')\n",
    "\n",
    "# # crc_8_CMD\n",
    "# data_mat, meta_data = preprocess_data_phyloseq(f'{overall_path}/pruned_crc_8_CMD', f'{overall_path}/cleaned_data/crc_8_CMD/crc_8_CMD', id = 'Sam_id', covar_l = [])\n",
    "# data_mat, meta_data = load_results_from_benchmarked_methods(f'{overall_path}/cleaned_data/crc_8_CMD/crc_8_CMD_count_data.csv', f'{overall_path}/cleaned_data/crc_8_CMD/crc_8_CMD_meta_data.csv')\n",
    "# check_complete_confounding(meta_data, \"study_name\", \"disease\", f'{overall_path}/cleaned_data/crc_8_CMD/crc_8_CMD')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "555b2b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_path = '/Users/hehehe/Desktop/CRC'\n",
    "data_mat, meta_data = preprocess_data_phyloseq(f'{overall_path}/MOSAIC/microbiomeHD Pruned', \n",
    "                                               f'{overall_path}/MOSAIC/cleaned_data/microbiomeHD/microbiomeHD', \n",
    "                                               id = 'Sam_id', \n",
    "                                               covar_l = [], \n",
    "                                               relab = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
